pkg/handlers/service_logs.go

```diff
@@ -4,88 +4,157 @@ import (
 	"encoding/json"
 	"fmt"
 	"net/http"
+	"regexp"
 
 	sdk "github.com/openshift-online/ocm-sdk-go"
 	slv1 "github.com/openshift-online/ocm-sdk-go/servicelogs/v1"
 	"github.com/openshift/ocm-agent-operator/api/v1alpha1"
 	"github.com/openshift/ocm-agent/pkg/consts"
-	log "github.com/sirupsen/logrus"
+	"github.com/prometheus/alertmanager/template"
 )
 
-type OCMClient interface {
-	SendServiceLog(summary, firingDesc, resolveDesc, clusterID string, severity v1alpha1.NotificationSeverity, logType string, references []v1alpha1.NotificationReferenceType, firing bool) error
+const (
+	ServiceLogActivePrefix  = "Issue Notification"
+	ServiceLogResolvePrefix = "Issue Resolution"
+)
+
+type ServiceLogBuilder struct {
+	wrappedBuilder *slv1.LogEntryBuilder
+	summary        string
+	firingDesc     string
+	resolveDesc    string
+	references     []v1alpha1.NotificationReferenceType
 }
 
-// ServiceLogsHandler manages service logs via the OCM API.
-type ServiceLogsHandler struct {
-	ocm *sdk.Connection
+type ServiceLog = slv1.LogEntry
+
+func NewServiceLogBuilder(summary, firingDesc, resolveDesc, clusterUUID string, severity v1alpha1.NotificationSeverity, logType string, references []v1alpha1.NotificationReferenceType) *ServiceLogBuilder {
+	return &ServiceLogBuilder{
+		wrappedBuilder: slv1.NewLogEntry().
+			ServiceName(consts.ServiceLogServiceName).
+			ClusterUUID(clusterUUID).
+			InternalOnly(false).
+			Severity(slv1.Severity(severity)).
+			LogType(slv1.LogType(logType)),
+		summary:     summary,
+		firingDesc:  firingDesc,
+		resolveDesc: resolveDesc,
+		references:  references,
+	}
 }
 
-// NewServiceLogsHandler creates a new handler for service logs.
-func NewServiceLogsHandler(o *sdk.Connection) *ServiceLogsHandler {
-	log.Debug("Creating new service logs handler")
-	return &ServiceLogsHandler{
-		ocm: o,
+var (
+	slVarRefRe = regexp.MustCompile(`\${[^{}]*}`)
+)
+
+func replacePlaceHoldersInString(s string, alert *template.Alert) (string, error) {
```

Comment:
Is it only me who feels the function is hard to understand?...

Reply:
This function is indeed a bit hard to understand.
I could add some comment telling that the place holders in `s` are replaced with the given `alert` labels & annotations.

The body of the function is also a bit hard to read because of the inner function but the code is actually much shorter that way. Explicitly looping on the matched token is actually clumsier; especially if we want to avoid recursive replacements (and we want that).